zone/parcels.cpp

```diff
@@ -0,0 +1,755 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; version 2 of the License.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY except by those people which sell it, which
+	are required to give you total support for your newly bought product;
+	without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+	A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with this program; if not, write to the Free Software
+	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+*/
+
+#include "../common/global_define.h"
+#include "../common/events/player_event_logs.h"
+#include "../common/repositories/trader_repository.h"
+#include "../common/repositories/character_parcels_repository.h"
+#include "worldserver.h"
+#include "string_ids.h"
+#include "parcels.h"
+#include "client.h"
+#include "../common/ruletypes.h"
+
+extern WorldServer worldserver;
+
+void Client::SendBulkParcels()
+{
+	SetEngagedWithParcelMerchant(true);
+	LoadParcels();
+
+	if (m_parcels.empty()) {
+		return;
+	}
+
+	ParcelMessaging_Struct pms{};
+	pms.packet_type = ItemPacketParcel;
+
+	std::stringstream           ss;
+	cereal::BinaryOutputArchive ar(ss);
+
+	for (auto &p: m_parcels) {
+		auto item = database.GetItem(p.second.item_id);
+		if (item) {
+			std::unique_ptr<EQ::ItemInstance> inst(database.CreateItem(item, p.second.quantity));
+			if (inst) {
+				inst->SetCharges(p.second.quantity > 0 ? p.second.quantity : 1);
+				inst->SetMerchantCount(1);
+				inst->SetMerchantSlot(p.second.slot_id);
+				if (inst->IsStackable()) {
+					inst->SetCharges(p.second.quantity);
+				}
+
+				if (item->ID == PARCEL_MONEY_ITEM_ID) {
+					inst->SetPrice(p.second.quantity);
+					inst->SetCharges(1);
+				}
+
+				pms.player_name     = p.second.from_name;
+				pms.sent_time       = p.second.sent_date;
+				pms.note            = p.second.note;
+				pms.serialized_item = inst->Serialize(p.second.slot_id);
+				pms.slot_id         = p.second.slot_id;
+				ar(pms);
+
+				uint32 packet_size = ss.str().length();
+				std::unique_ptr<EQApplicationPacket> out(new EQApplicationPacket(OP_ItemPacket, packet_size));
+				if (out->size != packet_size) {
+					LogError(
+						"Attempted to send a parcel packet of mismatched size {} with a buffer size of {}.",
+						out->Size(),
+						packet_size
+					);
+					return;
+				}
+				memcpy(out->pBuffer, ss.str().data(), out->size);
+				QueuePacket(out.get());
+
+				ss.str("");
+				ss.clear();
+			}
+		}
+
+	}
+	if (m_parcels.size() >= PARCEL_LIMIT) {
+		LogError(
+			"Found {} parcels for Character {}.  List truncated to the PARCEL_MAX_ITEMS [{}] + 5.",
+			m_parcels.size(),
+			GetCleanName(),
+			PARCEL_MAX_ITEMS
+		);
+		SendParcelStatus();
+		return;
+	}
+}
+
+void Client::SendParcel(Parcel_Struct parcel_in)
```

Comment:
Change this to `const Parcel_Struct&` to avoid the copy

Reply:
Update and also checked others and updated as well.